.github/actions/branch-lock/src/main.ts

```diff
@@ -0,0 +1,100 @@
+import * as core from '@actions/core'
+import {actionInputs} from 'github-actions-utils'
+import {Octokit} from '@octokit/core'
+
+/**
+ * GitHub Actions entrypoint.
+ */
+export async function run(): Promise<void> {
+  try {
+    // Validate inputs
+    const lock = actionInputs.getBool('lock', true)
+    const branch = actionInputs.getString('branch', true)
+    const token = actionInputs.getString('token', true, true)
+    const owner = actionInputs.getString('owner', true)
+    const repo = actionInputs.getString('repo', true)
+
+    // Create github service
+    const githubService = new GitHubService(token)
+    const branchId = await githubService.searchBranchProtection(owner, repo, branch)
+    await githubService.setLock(branchId, lock)
+    core.info('Succeed to update branch lock.')
+  } catch (err: unknown) {
+    if (err instanceof Error) {
+      core.setFailed(err)
+    } else {
+      core.setFailed('Unhandled error occured')
+    }
+  }
+}
+
+/**
+ * GitHub service to interact with GitHub GraphQL endpoint.
+ */
+class GitHubService {
+  // GraphQL client
+  private client: Octokit
+
+  /**
+   * Init github service.
+   * @param token github token.
+   */
+  constructor(token: string) {
+    this.client = new Octokit({auth: token})
+  }
+
+  async searchBranchProtection(owner: string, repo: string, branch: string): Promise<string> {
+    const query = `
+      query searchBranchProtection($repo: String!, $owner: String!) {
+        repository(name: $repo, owner: $owner) {
+          branchProtectionRules(first: 100) {
+            nodes {
+              id
+              pattern
+            }
+          }
+        }
+      }
+    `
+
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    const data: any = await this.client.graphql(query, {repo, owner})
+    for (const node of data.repository.branchProtectionRules.nodes) {
+      const pattern = new RegExp(node.pattern)
+      if (pattern.test(branch)) {
+        return node.id
+      }
+    }
+    throw new BranchNotFound(branch)
+  }
+
+  /**
+   * Set lock on branch protections.
+   * @param branchId branch id ref.
+   * @param lock lock or unlock branch.
+   */
+  async setLock(branchId: string, lock: boolean): Promise<void> {
+    const query = `
+      mutation setLock($branchId: ID!, $lock: Boolean!) {
+        updateBranchProtectionRule(input: {branchProtectionRuleId: $branchId, lockBranch: $lock}) {
+          clientMutationId
+        }
+      }
+    `
+
+    await this.client.graphql(query, {branchId, lock})
+  }
+}
+
+/**
+ * Represents a branch not found error.
+ */
+class BranchNotFound extends Error {
+  /**
+   * Init branch not found error.
+   * @param branchName branch name.
+   */
+  constructor(branchName: string) {
+    super(`Failed to find the branch: ${branchName}`)
+  }
+}
```

Comment:
I did not test it but this core replace your implementation

```suggestion
    var branchProtection = octokit.rest.repos.getBranchProtection({
      owner,
      repo,
      branch,
    });
    branchProtection.lockBranch=lock
    octokit.rest.repos.updateBranchProtection({
        owner,
        repo,
        branch,
        ...branchProtection
      })
    core.info('Succeed to update branch lock.')
  } catch (err: unknown) {
    if (err instanceof Error) {
      core.setFailed(err)
    } else {
      core.setFailed('Unhandled error occured')
    }
  }
}
```

Reply:
Unfortunately, the branch protection response from octokit doesn't match the request body in the updateBranchProtection method.
Assuming we install `octokit/rest` instead of `octokit/core`.

```typescript
    // Create github service
    const githubService = new Octokit({auth: token})
    const response = await githubService.rest.repos.getBranchProtection({
      owner,
      repo,
      branch,
    })
    const branchProtection = response.data
    branchProtection.lock_branch = {enabled: lock}
    await githubService.rest.repos.updateBranchProtection({
      owner,
      repo,
      branch,
      ...branchProtection
    })
```

The above implementation will lead to a compile error at build time.
```txt
src/main.ts:26:59 - error TS2345: Argument of type '{ url?: string | undefined; enabled?: boolean | undefined; required_status_checks?: { url?: string | undefined; enforcement_level?: string | undefined; contexts: string[]; checks: { context: string; app_id: number | null; }[]; contexts_url?: string | undefined; strict?: boolean | undefined; } | undefined; ... 15 mor...' is not assignable to parameter of type 'RequestParameters & Omit<ToOctokitParameters<{ parameters: { path: { owner: string; repo: string; branch: string; }; }; requestBody: { content: { "application/json": { required_status_checks: { ...; } | null; ... 9 more ...; allow_fork_syncing?: boolean | undefined; }; }; }; responses: { ...; }; }>, "baseUrl" | ... ...'.
  Type '{ url?: string | undefined; enabled?: boolean | undefined; required_status_checks?: { url?: string | undefined; enforcement_level?: string | undefined; contexts: string[]; checks: { context: string; app_id: number | null; }[]; contexts_url?: string | undefined; strict?: boolean | undefined; } | undefined; ... 15 mor...' is not assignable to type 'Omit<ToOctokitParameters<{ parameters: { path: { owner: string; repo: string; branch: string; }; }; requestBody: { content: { "application/json": { required_status_checks: { strict: boolean; contexts: string[]; checks?: { context: string; app_id?: number | undefined; }[] | undefined; } | null; ... 9 more ...; allow_...'.
    Types of property 'required_linear_history' are incompatible.
      Type '{ enabled?: boolean | undefined; } | undefined' is not assignable to type 'boolean | undefined'.
        Type '{ enabled?: boolean | undefined; }' is not assignable to type 'boolean | undefined'.
```

We need to pass most of the parameters explicitly.

Reply:
that part I lest to you I give you the API call

Reply:
Sure, I didn't move further yet.
See https://github.com/elastic/apm-pipeline-library/pull/2432#issuecomment-1803732747.
I prefer to wait until we agree on the solution before further development.